### PR TITLE
Update list-patterns.md

### PR DESCRIPTION
**Pull request by alrz**
Record recent LDM decisions and sync with the current impl.

Linking to test plan: https://github.com/dotnet/roslyn/issues/51289

### CODE REVIEW

File: proposals/list-patterns.md [hunk 1]
```diff
@@ -153,7 +148,7 @@ case {.., 1, _}: // expr.Length is >= 2 && expr[^2] is 1
 
 The order in which subpatterns are matched at runtime is unspecified, and a failed match may not attempt to match all subpatterns.
 
-> **Open question**: The pattern `{..}` tests for  `expr.Length >= 0`. Should we omit such test (assuming `Length` is always non-negative)?
+> **Open question**: By this definition, the pattern `{..}` tests for `expr.Length >= 0`. Should we omit such test, assuming `Length` is always non-negative? (answer [LDM 2021-05-26]: `{ .. }` will not emit a Length check)
```
Comment (alrz) [inline proposals/list-patterns.md:151]: :memo: Notes were clear that this is about semantics and not an optimization.

**Comment by alrz** [comment]
@333fred @jcouv for review. thanks.

(I didn't touch enumerable sections but I think we should move it to its own document eventually)

File: proposals/list-patterns.md [hunk 2]
```diff
@@ -69,28 +69,23 @@ Notes:
...
+A *list_pattern* is compatible with any type that is *countable* as well as *indexable* — it has an accessible indexer that takes an `Index` as the argument or otherwise an accessible indexer with a single `int` parameter. If both indexers are present, the former is preferred.  
 A *list_pattern* is also compatible with any type that is *enumerable*.
 
-A *slice_pattern* is compatible with any type that is *countable* as well as *sliceable* — it has an accessible indexer that takes a `Range` argument or otherwise an accessible `Slice` method that takes two `int` arguments. If both are present, the former is preferred.  
-A *slice_pattern* without a subpattern is also compatible with any type that is *enumerable*.
+A *slice_pattern* with a subpattern is compatible with any type that is *countable* as well as *sliceable* — it has an accessible indexer that takes a `Range` argument or otherwise an accessible `Slice` method with two `int` parameters. If both are present, the former is preferred.
+A *slice_pattern* without a subpattern is compatible with any type that is compatible with a *list_pattern*.
```
Comment (alrz) [inline proposals/list-patterns.md:81]: :memo: We don't require Slice method if there is no subpattern. This is now consistent with the decision about `{..}`.

File: proposals/list-patterns.md [hunk 3]
```diff
@@ -43,7 +43,7 @@ property_pattern
   ;
 
 slice_pattern
-  : '..' negated_pattern?
+  : '..' pattern?
```
Comment (alrz) [inline proposals/list-patterns.md:46]: Wasn't mentioned in LDM. But I think "all patterns" includes combinators.

Comment (333fred) [reply proposals/list-patterns.md:46]: It does. I'm not familiar enough with the various levels of the pattern grammar, does `negated_pattern` not include combinators?

File: proposals/list-patterns.md [hunk 4]
```diff
@@ -69,28 +69,23 @@ Notes:
...
+A *list_pattern* is compatible with any type that is *countable* as well as *indexable* — it has an accessible indexer that takes an `Index` as an argument or otherwise an accessible indexer with a single `int` parameter. If both indexers are present, the former is preferred.  
 A *list_pattern* is also compatible with any type that is *enumerable*.
 
-A *slice_pattern* is compatible with any type that is *countable* as well as *sliceable* — it has an accessible indexer that takes a `Range` argument or otherwise an accessible `Slice` method that takes two `int` arguments. If both are present, the former is preferred.  
-A *slice_pattern* without a subpattern is also compatible with any type that is *enumerable*.
+A *slice_pattern* with a subpattern is compatible with any type that is *countable* as well as *sliceable* — it has an accessible indexer that takes a `Range` as an argument or otherwise an accessible `Slice` method with two `int` parameters. If both are present, the former is preferred.
+A *slice_pattern* without a subpattern is compatible with any type that is compatible with a *list_pattern*.
```
Comment (333fred) [inline proposals/list-patterns.md:81]: Might need to mention discard subpattern?

Comment (alrz) [reply proposals/list-patterns.md:81]: With this change, we have:

- `{..}` works even without `this[Range]` or a proper `Slice` method.
- `{.._}` requires `this[Range]`/`Slice`. The codegen, however, is permitted to omit the invocation (which we do). That is not limited to discards or the slice pattern, we skip an evaluation when the output is unused, e.g. with an unused variable etc.

Comment (alrz) [reply proposals/list-patterns.md:81]: Sounds like we currently don't include unused variables here, but it's something that could be done as an optimization.

Comment (alrz) [reply proposals/list-patterns.md:81]: Here's an example to demonstrate it's not limited to discard: `if (this is {IntProp: {}}){}` IntProp is not called here.

File: proposals/list-patterns.md (hunk 3 above)
Comment (alrz) [reply proposals/list-patterns.md:46]: No it doesn't. I've started this as an unary operator, hence the precedence.

File: proposals/list-patterns.md (hunk 4 above)
Comment (333fred) [reply proposals/list-patterns.md:81]: > The codegen, however, is permitted to omit the invocation (which we do).

This is a good distinction, but I do think it is good to make this not an optimization: it gets back to making sure that behavior of something like `default(ImmutableArray<int>) is { .._ }` is well-defined.

Comment (alrz) [reply proposals/list-patterns.md:81]: At the moment the behavior of `default(ImmutableArray<int>) is { .._ }` is an error. "unsupported type for slice pattern"

ImmutableArray doesn't expose a Slice method or range indexer.

Comment (333fred) [reply proposals/list-patterns.md:81]: Yeah, but additions are tracked by https://github.com/dotnet/runtime/issues/22928.

Comment (alrz) [reply proposals/list-patterns.md:81]: My main concern is that when a type is not sliceable, we should be able to do `{1, ..}` etc. Not every list should be required to be sliceable only to be able to skip elements in the list pattern.

Comment (jcouv) [reply proposals/list-patterns.md:81]: I'm fine with the behavior as spec'ed in this PR. I'd let `.._` and `..var unused` require a sliceable type. It's conceptually simpler.

For `{ ..var unused }`: during binding we don't know that `var unused` is unused.
For `{ .._ }`: @333fred , I can add a note to test plan to rediscuss if you think this should not require a sliceable type.

**Comment by alrz** [comment]
To be clear, I don't think we should do anything specific for `ImmutableArray`, it's not a known type in the compiler and I don't believe it should influence the design. However, it's a use case that we definitely want to cover in this area.

To that end, I've proposed to use a hypothetical `bool Deconstruct` (already championed in some form) so that we can safely match an ImmutableArray whether or not it's initialized. I'm not sure if we want to go down that road though. But checking `IsDefault` is too type specific IMO, as far as I know, that doesn't have any special meaning anywhere in the language or BCL in general.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Comment by jcouv** [comment]
> I don't think we should do anything specific for ImmutableArray,

Agreed.
FYI, based on email discussion, I suspect there will be a formal proposal soon to add a `null/default` operator. But that'll be orthogonal to the list-patterns feature.

File: proposals/list-patterns.md (hunk 3 above)
Comment (jcouv) [reply proposals/list-patterns.md:46]: nit: Consider adding links to previous patterns specs:

https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.0/pattern-matching.md
https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/patterns.md
https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/patterns3.md

File: proposals/list-patterns.md (hunk 4 above)
Comment (alrz) [reply proposals/list-patterns.md:81]: >For `{ .._ }`

Just like we still need an indexable type for `{_}` that we're not going to use, `{.._}` should require a sliceable type that we may not use.  I think the part that we decide to not evaluate the method, is orthogonal to *binding rules*.

**Comment by jcouv** [comment]
@333fred I started thread with LDM mentioning the `{ .._ }` question. Feel free to chime in. I think we can merge this PR if you're good with it (I am).

**Review by 333fred — APPROVED** [review]

**Comment by jcouv** [comment]
Merged/squashed. Thanks @alrz

**Pull request merged**